Adding an elasticache snapshot module.

s-hertel opened the pull request:
Allows user to create, copy, or delete a snapshot.

##### ISSUE TYPE
 - New Module Pull Request

##### COMPONENT NAME
lib/ansible/modules/cloud/amazon/elasticache_snapshot.py

##### ANSIBLE VERSION
```
ansible 2.3.0 (elasticache-snapshot 1dba88717d) last updated 2017/02/07 17:31:11 (GMT -400)
  config file =
  configured module search path = Default w/o overrides
```

##### SUMMARY
Create, delete, and copy snapshots with elasticache. #20842 showed interest in improved functionality of ElastiCache modules.

Test playbook:
```
---
- hosts: localhost
  connection: local
  tasks:
    - name: "create snapshot"
      elasticache_snapshot:
        name: "test-snapshot"
        replication_id: "{{ replication }}"
        cluster_id: "{{ cluster }}"
        state: "present"
        region: "{{ region }}"
        profile: "{{ user }}"
```
Output:
```
Loading callback plugin default of type stdout, v2.0 from /Users/shertel/Workspace/ansible/lib/ansible/plugins/callback/__init__.py

PLAYBOOK: ec_create_snapshot.yml *********************************************************
1 plays in my_playbooks/ec_create_snapshot.yml

PLAY [localhost] ******************************************************************

TASK [Gathering Facts] ************************************************************
Using module file /Users/shertel/Workspace/ansible/lib/ansible/modules/system/setup.py
<127.0.0.1> ESTABLISH LOCAL CONNECTION FOR USER: shertel
<127.0.0.1> EXEC /bin/sh -c 'echo ~ && sleep 0'
<127.0.0.1> EXEC /bin/sh -c '( umask 77 && mkdir -p "` echo /Users/shertel/.ansible/tmp/ansible-tmp-1486572362.830099-86518620017685 `" && echo ansible-tmp-1486572362.830099-86518620017685="` echo /Users/shertel/.ansible/tmp/ansible-tmp-1486572362.830099-86518620017685 `" ) && sleep 0'
<127.0.0.1> PUT /var/folders/by/k8_fbl593dlctgqmwq5wzl2c0000gn/T/tmp7ickqs3y TO /Users/shertel/.ansible/tmp/ansible-tmp-1486572362.830099-86518620017685/setup.py
<127.0.0.1> EXEC /bin/sh -c 'chmod u+x /Users/shertel/.ansible/tmp/ansible-tmp-1486572362.830099-86518620017685/ /Users/shertel/.ansible/tmp/ansible-tmp-1486572362.830099-86518620017685/setup.py && sleep 0'
<127.0.0.1> EXEC /bin/sh -c '/Users/shertel/Workspace/ansible/venv/python3env/bin/python /Users/shertel/.ansible/tmp/ansible-tmp-1486572362.830099-86518620017685/setup.py; rm -rf "/Users/shertel/.ansible/tmp/ansible-tmp-1486572362.830099-86518620017685/" > /dev/null 2>&1 && sleep 0'
ok: [localhost]
META: ran handlers

TASK [create snapshot] ************************************************************
task path: /Users/shertel/Workspace/ansible/my_playbooks/ec_create_snapshot.yml:5
Using module file /Users/shertel/Workspace/ansible/lib/ansible/modules/cloud/amazon/elasticache_snapshot.py
<127.0.0.1> ESTABLISH LOCAL CONNECTION FOR USER: shertel
<127.0.0.1> EXEC /bin/sh -c 'echo ~ && sleep 0'
<127.0.0.1> EXEC /bin/sh -c '( umask 77 && mkdir -p "` echo /Users/shertel/.ansible/tmp/ansible-tmp-1486572368.488385-166467240966919 `" && echo ansible-tmp-1486572368.488385-166467240966919="` echo /Users/shertel/.ansible/tmp/ansible-tmp-1486572368.488385-166467240966919 `" ) && sleep 0'
<127.0.0.1> PUT /var/folders/by/k8_fbl593dlctgqmwq5wzl2c0000gn/T/tmp21s1_0z3 TO /Users/shertel/.ansible/tmp/ansible-tmp-1486572368.488385-166467240966919/elasticache_snapshot.py
<127.0.0.1> EXEC /bin/sh -c 'chmod u+x /Users/shertel/.ansible/tmp/ansible-tmp-1486572368.488385-166467240966919/ /Users/shertel/.ansible/tmp/ansible-tmp-1486572368.488385-166467240966919/elasticache_snapshot.py && sleep 0'
<127.0.0.1> EXEC /bin/sh -c '/Users/shertel/Workspace/ansible/venv/python3env/bin/python /Users/shertel/.ansible/tmp/ansible-tmp-1486572368.488385-166467240966919/elasticache_snapshot.py; rm -rf "/Users/shertel/.ansible/tmp/ansible-tmp-1486572368.488385-166467240966919/" > /dev/null 2>&1 && sleep 0'
changed: [localhost] => {
    "changed": true,
    "invocation": {
        "module_args": {
            "aws_access_key": null,
            "aws_secret_key": null,
            "bucket": null,
            "cluster_id": "new",
            "ec2_url": null,
            "name": "test-snapshot",
            "profile": "shertel",
            "region": "us-east-1",
            "replication_id": "test-please-delete",
            "security_token": null,
            "state": "present",
            "target": null,
            "validate_certs": true
        }
    },
    "response_metadata": {
        "http_headers": {
            "content-length": "1478",
            "content-type": "text/xml",
            "date": "Wed, 08 Feb 2017 16:45:43 GMT",
            "x-amzn-requestid": "08c55b18-ee1e-11e6-879f-437db62c7900"
        },
        "http_status_code": 200,
        "request_id": "08c55b18-ee1e-11e6-879f-437db62c7900",
        "retry_attempts": 0
    },
    "snapshot": {
        "auto_minor_version_upgrade": true,
        "cache_cluster_create_time": "2017-02-08T16:03:14.589000+00:00",
        "cache_cluster_id": "new",
        "cache_node_type": "cache.m1.small",
        "cache_parameter_group_name": "default.redis3.2",
        "cache_subnet_group_name": "default",
        "engine": "redis",
        "engine_version": "3.2.4",
        "node_snapshots": [
            {
                "cache_node_create_time": "2017-02-08T16:03:14.589000+00:00",
                "cache_node_id": "0001",
                "cache_size": ""
            }
        ],
        "num_cache_nodes": 1,
        "port": 11211,
        "preferred_availability_zone": "us-east-1d",
        "preferred_maintenance_window": "wed:03:00-wed:04:00",
        "snapshot_name": "test-snapshot",
        "snapshot_retention_limit": 0,
        "snapshot_source": "manual",
        "snapshot_status": "creating",
        "snapshot_window": "10:00-11:00",
        "vpc_id": "vpc-c248fda4"
    }
}
META: ran handlers
META: ran handlers

PLAY RECAP ************************************************************************
localhost                  : ok=2    changed=1    unreachable=0    failed=0
```

s-hertel committed:
Adding an elasticache snapshot module.

Allows user to create, copy, or delete a snapshot.

s-hertel committed:
Removing unnecessary function

s-hertel committed:
Make indentation uniform.

ryansb commented on lib/ansible/modules/cloud/amazon/elasticache_snapshot.py:
```diff
@@ -0,0 +1,236 @@
+
+DOCUMENTATION = """
+---
+module: elasticache
```
Name needs to be updated here.

ryansb commented on lib/ansible/modules/cloud/amazon/elasticache_snapshot.py:
```diff
@@ -0,0 +1,236 @@
+"""
+
+RETURN = """
+elasticache:
```
Instead of nesting under the `elasticache` key, you can pass the info back as `exit_json(changed=[whatever], **camel_dict_to_snake_dict(response))`. That would change usage from:

```
- elasticache_snapshot: [args]
  register: snapshot_info
- debug: var=snapshot_info.elasticache.[things]

# to this
- debug: var=snapshot_info.[things]
```

For facts modules, it makes sense to nest the returns, as those are returned and automatically entered to the namespace without the user `register`ing them.

ryansb commented on lib/ansible/modules/cloud/amazon/elasticache_snapshot.py:
```diff
@@ -0,0 +1,236 @@
+    response    = {}
+
+    if state == 'present':
+        for req in [replication_id, cluster_id]:
```
This can be changed to
```
if not all((replication_id, cluster_id)):
    module.fail_json(....)
```

ryansb commented on lib/ansible/modules/cloud/amazon/elasticache_snapshot.py:
```diff
@@ -0,0 +1,236 @@
+    elif state == 'absent':
+        response, changed = delete(module, connection, name)
+    elif state == 'copy':
+        for req in [target, bucket]:
```
Same as line 215.

ryansb commented on lib/ansible/modules/cloud/amazon/elasticache_snapshot.py:
```diff
@@ -0,0 +1,236 @@
+            if not req:
+                module.fail_json(msg="The state 'copy' requires options: 'target' and 'bucket'.")
+        response, changed = copy(module, connection, name, target, bucket)
+    
```
This line has trailing whitespace, and is causing the shippable PEP8 checks to fail.

ryansb commented on lib/ansible/modules/cloud/amazon/elasticache_snapshot.py:
```diff
@@ -0,0 +1,236 @@
+        bucket          = dict(required=False, type='str'),
+    )
+    )
+        
```
Trailing whitespace on a blank line is failing the PEP8 checks.

s-hertel committed:
Making requested changes.

Fixing PEP8

Adding a more graceful fail for delete() if the snapshot's state valid (such as when it is in the process of being created).

s-hertel committed:
PEP8

ryansb commented on lib/ansible/modules/cloud/amazon/elasticache_snapshot.py:
```diff
@@ -174,7 +174,8 @@ def delete(module, connection, name):
             changed = False
         elif e.response['Error']['Code'] == "InvalidSnapshotState":
-            module.fail_json(msg="Error: InvalidSnapshotState. The snapshot is not in an available state or failed state to allow deletion. You may need to wait a few minutes.")
+            module.fail_json(msg="Error: InvalidSnapshotState. The snapshot is not in an available state or failed state to allow deletion."
```
This just joins the two strings, so you need a trailing space on this line so it doesn't come out as `...allow deletion.You may need...`

s-hertel committed:
Fixing some formatting

move imports

fix parameter alignment

willthames commented on lib/ansible/modules/cloud/amazon/elasticache_snapshot.py:
```diff
@@ -234,7 +231,8 @@ def main():
 # import module snippets
 from ansible.module_utils.basic import *
-from ansible.module_utils.ec2 import *
+from ansible.module_utils.ec2 import boto3_conn, get_aws_connection_info, ec2_argument_spec, camel_dict_to_snake_dict
```
These should be towards the top, just below the end of the documentation sections

willthames commented on lib/ansible/modules/cloud/amazon/elasticache_snapshot.py:
```diff
@@ -234,7 +231,8 @@ def main():
 
 # import module snippets
 from ansible.module_utils.basic import *
```
This should be `from ansible.module_utils.basic import AnsibleModule`

s-hertel committed:
move imports to the top of the file below documentation

willthames commented:
shipit

tomislacker commented on lib/ansible/modules/cloud/amazon/elasticache_snapshot.py:
```diff
@@ -0,0 +1,238 @@
+    response = {}
+
+    if state == 'present':
+        if not all((replication_id, cluster_id)):
```
@s-hertel I believe this is actually incorrect. I must omit the `cluster_id` parameter to get the ElastiCache service to take a snapshot of a replication group. In the next few days, I'll get my changes put into a PR and I can show you. I've also added features for optionally waiting until a snapshot is in the "available" state.

s-hertel replied:
Oh. You're right, that's definitely not what's intended. PRs welcome! Is your cluster_id parameter == None?
